include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,431 @@
+      }
+    }
+    if (sync_stream_)
+      CUDA_DTOR_CALL(cudaStreamSynchronize(sync_stream_));
+  }
+
+ private:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    return allocate_from_global_pool(bytes, alignment);
+  }
+
+  void do_deallocate(void *mem, size_t bytes, size_t alignment) override {
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    char *ptr = static_cast<char *>(mem);
+    pop_block_padding(ptr, bytes, alignment);
+    return global_pool_.deallocate(ptr, bytes, alignment);
+  }
+
+  /**
+   * @brief Tries to recycle per-stream free memory or allocates from the global pool.
+   *
+   * There are four cases:
+   * (1) Attempt to allocate from stream-specific free list - a smallest suitable block is used,
+   *     padding is applied, if necessary, and recorded.
+   * (2) If (1) fails and there are no pending per-stream free operations, the requested memory
+   *     is allocated from the global pool.
+   * (3) If (1) fails and there are free pending per-stream operations, an attempt is made to get
+   *     memory from the free list of the underlying global pool, without using upstream resource.
+   *     If this allocation succeeds, the memory allocated this way is returned.
+   * (4) If (3) fails, the pending per-stream frees are collected to the global pool and then
+   *     an allocation is made from the global pool.
+   *
+   * Per-stream frees are not coalesced - therefore, even if the total memory freed on a stream
+   * may be sufficient, there may be no satisfactory block.
+   */
+  void *do_allocate_async(size_t bytes, size_t alignment, stream_view stream) override {
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    auto it = stream_free_.find(stream.value());
+    void *ptr;
+    if (it != stream_free_.end()) {
+      ptr = try_allocate(it->second, bytes, alignment);
+      if (ptr)
+        return ptr;
+    }
+    return allocate_from_global_pool(bytes, alignment);
+  }
+
+  /**
+   * @brief Allocates from a global pool, possibly releasing per-stream memory to the global pool.
+   *
+   * If the global pool can't allocate enough memory without allocating upstream blocks,
+   * and there are some per-stream pending frees, the completed per-streem frees are returned
+   * to the global pool.
+   * Then, another request to the global pool is made, this time allowing allocation from upstream.
+   */
+  void *allocate_from_global_pool(size_t bytes, size_t alignment) {
+    void *ptr;
+    if (num_pending_frees_ == 0) {
+      ptr = global_pool_.allocate(bytes, alignment);
+      return ptr;
+    }
+    ptr = global_pool_.try_allocate_from_free(bytes, alignment);
+    if (ptr)
+      return ptr;
+    for (auto &kv : stream_free_)
```

Comment:
Maybe there should be 5), where you synchronize to make sure that all `PerStreamFreeBlocks` are ready and then allocate try again global_pool_?

Reply:
It would make sense if `upstream` is known to be slow (cudaMalloc, cudaHostAlloc). If `upstream` is not that slow (i.e. another pool/slab/) the performance would suffer. The idea itself is nice, but I'd make it configurable.

Reply:
It is just an idea.

Reply:
Done.